std/format.d

```diff
@@ -7017,3 +7063,538 @@ char[] sformat(Char, Args...)(return scope char[] buf, scope const(Char)[] fmt,
     assert(c == "^-nan         $", "\ngot:'"~ c ~ "'\nexp:'^-nan         $'");
     assert(d == "^-NAN         $", "\ngot:'"~ d ~ "'\nexp:'^-NAN         $'");
 }
+
+private auto printFloat(T, Char)(T val, FormatSpec!Char f, FloatingPointControl.RoundingMode rm)
+if (is(T == float) || is(T == double) || (is(T == real) && T.mant_dig == double.mant_dig))
+{
+    import std.conv : to;
+    import std.algorithm.comparison : max;
+
+    enum byte[16] alpha = ['0','1','2','3','4','5','6','7','8','9','a','b','c','d','e','f'];
+    enum byte[16] Alpha = ['0','1','2','3','4','5','6','7','8','9','A','B','C','D','E','F'];
+
+    union FloatBits
+    {
+        T floatValue;
+        ulong ulongValue;
+    }
+
+    FloatBits fb;
+    fb.floatValue = val;
+    ulong ival = fb.ulongValue;
+
+    static if (is(T == float))
+    {
+        ulong mnt = ival & ((1L << 23) - 1);
+        int exp = (ival >> 23) & ((1L << 8) - 1);
+        int maxexp = 255;
+        string sgn = (ival >> (23 + 8)) & 1 ? "-" : "";
+        enum long bias = 127;
+    }
+    else
+    {
+        version (DigitalMars)
+        {
+            // hack to work around issue 20363
+            ival ^= rm;
+            ival ^= rm;
+        }
+
+        ulong mnt = ival & ((1L << 52) - 1);
+        int exp = (ival >> 52) & ((1L << 11) - 1);
+        int maxexp = 2047;
+        string sgn = (ival >> (52 + 11)) & 1 ? "-" : "";
+        enum long bias = 1023;
+    }
+
+    if (sgn == "" && f.flPlus) sgn = "+";
+    if (sgn == "" && f.flSpace) sgn = " ";
+
+    assert(f.spec == 'a' || f.spec == 'A');
+    bool is_upper = f.spec == 'A';
+
+    // special treatment for nan and inf
+    if (exp == maxexp)
+    {
+        char[] result;
+        result.length = max(f.width, sgn.length + 3);
+        result[] = ' ';
+
+        auto offset = f.flDash ? 0 : (result.length - 3);
+
+        if (sgn!="")
+        {
+            if (f.flDash) ++offset;
+            result[offset-1] = sgn[0];
+        }
+
+        if (mnt == 0)
+        {
+            result[offset + 0] = is_upper?'I':'i';
+            result[offset + 1] = is_upper?'N':'n';
+            result[offset + 2] = is_upper?'F':'f';
+        }
+        else
+        {
+            result[offset + 0] = is_upper?'N':'n';
+            result[offset + 1] = is_upper?'A':'a';
+            result[offset + 2] = is_upper?'N':'n';
+        }
```

Comment:
please write this as 
```
result[offset .. offset +3] = (mnt == 0) ? ( is_upper ? "INF" : "inf" )
                                                                  : ( is_upper ? "NAN" : "nan" )
```

Reply:
When I remember right, it were speed considerations. But I don't know, if printing NaNs need to be fast...

Reply:
Indeed. Which compiler did you use? I could imagine DMD doing a terrible job optimising that but LDC and GDC should have no problem.

Reply:
It's too long ago, I do not remember. I did a lot of checks with dmd and ldc2 with and without optimization. From different tests it seems, that `result[a..b] = something` is generally slower than expected, even with optimization.

Reply:
Hmm, I would have thought that the LDC simplify druntime pass would have dealt with that fine. maybe the conditionals throw it off? 
Still, I don't think that too many people's programs will be slowed down by writing NaNs, and if for some reason they are they can use LTO with druntime which should definitely work.

Reply:
I meanwhile ran all variants though the compilers and there is little difference.

Reply:
Good to see.